core-bundle/src/Resources/contao/library/Contao/Widget.php

```diff
@@ -1378,22 +1379,16 @@ public static function getAttributesFromDca($arrData, $strName, $varValue=null,
 				{
 					$value = $blnIsAssoc ? $kk : $vv;
 
-					if ($varValue && $varValue == $value)
+					if (($i = array_search($value, $unknown)) !== false)
 					{
-						$isKnownOption = true;
+						unset($unknown[$i]);
 					}
 
 					$arrAttributes['options'][$key][] = array('value'=>$value, 'label'=>($blnUseReference ? ((($ref = (\is_array($arrData['reference'][$vv]) ? $arrData['reference'][$vv][0] : $arrData['reference'][$vv])) != false) ? $ref : $vv) : $vv));
 				}
 			}
 
-			// If the value is not in the options array, the current user most
-			// likely cannot access it. We add the value as unknown option, so
-			// it does not get lost when saving the record (see #920).
-			if ($varValue && !$isKnownOption)
-			{
-				$arrAttributes['options'][] = array('value'=>$varValue, 'label'=>$GLOBALS['TL_LANG']['MSC']['unknownOption']);
-			}
+			$arrAttributes['unknownOption'] = array_filter($unknown);
```

Comment:
What is the purpose of `array_filter()` here?

Reply:
It filters empty entries, which can never be valid options.

Reply:
This would also strips values like `int 0` and `string '0'`. Is it supposed to be stripped even for arrays not being associative?

Reply:
Yes. `0` and `'0'` cannot be valid options in Contao 4.9. There is a PR to change this behavior in a future version, however, we fist have to make sure that there are no side-effects (see #2350).